### DISCUSSION TITLE
Some problems in crytatl.i material module, regarding crystal plasticity

### DISCUSSION THREAD

**Discussion by PengWei97** [idaholab/moose#17592] [Q&A Modules: Solid mechanics] [answered]
Dear MOOSE experts,

In order to complete the #17577  job, I have some doubts about the elastic modulus and crystal plastic constitutive part of the material in crystal.i. I hope to get your guidance. 

According to my understanding, `FiniteStrainCrystalPlasticity`is responsible for the constitutive part of crystal plasticity,in the materials module of [crystal.i](https://github.com/PengWei97/kunpeng/blob/main/test/tests/crystalPlastic_phaseField_GG/crysp.i). 

And `ComputeElasticityTensorCP `is responsible for inputting Euler angles and rotation matrix to the FiniteStrainCrystalPlasticity sub-module. I have the following questions and hope to get your answers,

First, does `FiniteStrainCrystalPlasticity `only need to input `_elasticity_tensor `and `ComputeElasticityTensorCP `from `ComputeElasticityTensorCP `at initial time? and these two variables are already included in the MaterialProperty variable at timestep_begin, I don’t know if my understanding is correct?

Second, the material rotation annotated in `MaterialProperty<RankTwoTensor>&_update_rot` refers to the rotation caused by the entire material deformation process during the loading process, that is, 

> rot = get_current_rotation(_deformation_gradient[_qp]); // Calculate material rotation
> _update_rot[_qp] = rot * _crysrot[_qp]; 

The crystal orientation refers to the rotation tensor, `_crysrot`, calculated after the material initially gives the crystal grain Euler angles. If my understanding is wrong, please give me more guidance.

Any suggestions or recommendations would be greatly appreciated.

Thank you
Wei Peng

**Comment by PengWei97** [comment]
@jiangwen84，Could you give me some guidance?

**Comment by jiangwen84** [comment, marked as answer]
> Dear MOOSE experts,
> 
> In order to complete the #17577 job, I have some doubts about the elastic modulus and crystal plastic constitutive part of the material in crystal.i. I hope to get your guidance.
> 
> According to my understanding, `FiniteStrainCrystalPlasticity`is responsible for the constitutive part of crystal plasticity,in the materials module of [crystal.i](https://github.com/PengWei97/kunpeng/blob/main/test/tests/crystalPlastic_phaseField_GG/crysp.i).
> 
> And `ComputeElasticityTensorCP `is responsible for inputting Euler angles and rotation matrix to the FiniteStrainCrystalPlasticity sub-module. I have the following questions and hope to get your answers,
> 
> First, does `FiniteStrainCrystalPlasticity `only need to input `_elasticity_tensor `and `ComputeElasticityTensorCP `from `ComputeElasticityTensorCP `at initial time? and these two variables are already included in the MaterialProperty variable at timestep_begin, I don’t know if my understanding is correct?

`_elasticity_tensor` is initial rotated elasticity tensor. If the Euler angles do not change, its value remains the same during  every linear residual calculation.  

> 
> Second, the material rotation annotated in `MaterialProperty<RankTwoTensor>&_update_rot` refers to the rotation caused by the entire material deformation process during the loading process, that is,
> 
> > rot = get_current_rotation(_deformation_gradient[_qp]); // Calculate material rotation
> > _update_rot[_qp] = rot * _crysrot[_qp];
> 
> The crystal orientation refers to the rotation tensor, `_crysrot`, calculated after the material initially gives the crystal grain Euler angles. If my understanding is wrong, please give me more guidance.
> 

`_crysrot` is the initial grain orientation. The `rot` is the incremental rotation due to deformation. 

> Any suggestions or recommendations would be greatly appreciated.
> 
> Thank you
> Wei Peng

**Reply by PengWei97** [reply]
Thank you very much @jiangwen84  for your patient reply. 

According to my understanding, `ComputeElasticityTensorCP` is responsible for outputting the elasticity modulus after rotation, namely `_elasticity_tensor` .

And the rotation matrix, namely `_crysrot` ,  for update slip direction and normal with crystal orientation.

Is this correct?

wei

**Reply by jiangwen84** [reply]
> Thank you very much @jiangwen84 for your patient reply.
> 
> According to my understanding, `ComputeElasticityTensorCP` is responsible for outputting the elasticity modulus after rotation, namely `_elasticity_tensor` .
> 

`ComputeElasticityTensorCP` calculates the elasticity tensor `_elasticity_tensor` based on grain orientation (Euler angles). The `_elasticity_tensor` will be used in `CrystalPlasticity Stress Calculation` to compute the stress at intermediate configuration. You can also visualize the component of `_elasticity_tensor` using a AuxVariable. 

> And the rotation matrix, namely `_crysrot` , for update slip direction and normal with crystal orientation.
> 

`_crysrot` is basically the rotation matrix for grain orientation. It is used to rotate the slip system. The rotation due to deformation is calculated by `_update_rotation`.   

> Is this correct?
> 
> wei

**Reply by PengWei97** [reply]
Thank you @jiangwen84  very much for your guidance. I have understood the role of two material parameters.

Please forgive me for repeatedly determining whether my understanding is correct, because I need to develop a model that couples crystal plasticity and phase field to study the mechanical field, especially the influence of plasticity on the behavior of grain growth. So I need to confirm that only two `MaterialProperty`s in `ComputeElasticityTensorCP`: `_elasticity_tensor` and `_crysrot` are output as the input parameters of `FiniteStrainCrystalPlasticity`.

Based on the existing coupling model of moose, it is the corresponding coupled linear elastic force field, and the file is [bicrystal.i](https://github.com/idaholab/moose/blob/06bc1f73f0d4dfb3d01ce9ba77d6bb97346e07eb/modules/combined/test/tests/ACGrGrElasticDrivingForce/bicrystal.i). The current thinking about crystal plasticity and phase-field coupled simulation is to replace the method of `Element property read user object` in `ComputeElasticityTensorCP` with `GrainDataTracker` to assign elastic modulus and Euler angles. I wonder if it is reasonable? I hope you can give me some suggestions.

Thanks for the help,
Best,
wei

**Reply by PengWei97** [reply]
hi @jiangwen84，

Thank you very much for your careful guidance. Through reading the tensor module on the moose official website, I understand that the module that calculates the elastic tensor needs to output two material parameters: elastic modulus and rotation matrix.

Finally, if you have any suggestions for issue #17577, your guidance is welcome.

Best,
wei
